OpenSim/Actuators/VandenBogert2011Muscle.cpp

```diff
@@ -0,0 +1,1234 @@
+
+    VandenBogert2011Muscle::ImplicitResidual opPoint;
+    opPoint=calcImplicitResidual (y,ydot,muscleLength,u,0);
+    double opForceResidual=opPoint.forceResidual;
+    double opActivResidual=opPoint.activResidual;
+
+    VandenBogert2011Muscle::ImplicitResidual del;
+
+    //----------df_dy------------//
+    SimTK::Vec2 dh = {h,0};
+    del = calcImplicitResidual(y+dh,ydot,muscleLength,u,0);
+    df_dy[0][0]= (del.forceResidual-opForceResidual)/h;
+    df_dy[1][0]= (del.activResidual-opActivResidual)/h;
+
+    dh = {0,h};
+    del = calcImplicitResidual(y+dh,ydot,muscleLength,u,0);
+    df_dy[1][0]= (del.forceResidual-opForceResidual)/h;
+    df_dy[1][1]= (del.activResidual-opActivResidual)/h;
+
+    //----------df_dydot------------//
+    dh = {h,0};
+    del = calcImplicitResidual(y,ydot+dh,muscleLength,u,0);
+    df_dydot[0][0]= (del.forceResidual-opForceResidual)/h;
+    df_dydot[1][0]= (del.activResidual-opActivResidual)/h;
+
+    dh = {0,h};
+    del = calcImplicitResidual(y,ydot+dh,muscleLength,u,0);
+    df_dydot[1][0]= (del.forceResidual-opForceResidual)/h;
+    df_dydot[1][1]= (del.activResidual-opActivResidual)/h;
+
+    //----------df_du----------------//
+    del = calcImplicitResidual(y,ydot,muscleLength,u+h,0);
+    df_du[0]= (del.forceResidual-opForceResidual)/h;
+    df_du[1]= (del.activResidual-opActivResidual)/h;
+
+/*
+    for (int i =0; i<=1; i=i+1) {
+        yTemp[i]=y[i]+h;
+        del = calcImplicitResidual(yTemp,ydot,u,0);
+        df_dy[0][i]=(del.forceResidual-opForceResidual)/h; //force residual
+        df_dy[1][i]=(del.activResidual-opActivResidual)/h; //activ residual
+
+
+        ydotTemp[i]=y[i]+h;
+        del = calcImplicitResidual(y,ydotTemp,u,0);
+
+        df_dydot[0][i]=(del.forceResidual-opForceResidual)/h; //force residual
+        df_dydot[1][i]=(del.activResidual-opActivResidual)/h; //activ residual
+
+        yTemp=y;
+        ydotTemp=ydot;
+
+    }*/
+
+
+    //del = calcImplicitResidual(y,ydot,u+h,0);
+    //double df_du = (del.forceResidual-opForceResidual)/h;
+
+    opPoint.df_dy=df_dy;
+    opPoint.df_dydot=df_dydot;
+    opPoint.df_du=df_du;
+
+    return opPoint;
+
+}
+
+
+//------------------------------------------------------------------------------
+SimTK::Vec3 VandenBogert2011Muscle::calcFiberStaticEquilibResidual(
+        double projFibLen, double muscleLength, double activ) const
+{
+// Just a quick convience method to get the force residual
+// under static conditions
+
+    VandenBogert2011Muscle::ImplicitResidual Results = calcImplicitResidual(
+            muscleLength, projFibLen, activ, 0.0, 0.0, activ, 1.0);
+
+    SimTK::Vec3 resAndDerivative;
+
+    resAndDerivative[0]=Results.forceResidual;
+    resAndDerivative[1]=Results.df_dy[0][0];
+    resAndDerivative[2]=Results.forceTendon;
+
+    return resAndDerivative;
+
+}
+
+//------------------------------------------------------------------------------
+SimTK::Vec2 VandenBogert2011Muscle::calcFiberStaticEquilbirum(
```

Comment:
There should be some high level description saying that this is a Newton-Raphson method, etc.